For discussion: Rework Aad package for null safety + flutter web support

stevenebutler opened the pull request:
This pull request is more for feedback/discussion. I have made some significant breaking changes to the library that may not be acceptable.

* flutter web support (using MSAL.js and package:js)
* support for null-safety (required use of flutter_webview, and meant architectural changes)
* added a dependency on bloc and flutter_bloc libraries to manage the use of flutter_webview
* introduced a token provider concept that has different implementations depending on mobile/web/stubbed

flutter_webview is the officially supported flutter webview widget, and it sits in the widget tree instead of overlaying it.
This means you need to figure out where to put your login screen instead of just placing it over the top of your normal
tree. A benefit to this is you no longer need to work out how to size/resize the webview as this happens automatically
through the normal flutter rebuild process. 

One problem I've found with using flutter_webview is that the webview seems to be hidden from features like the android password manager. Therefore, I don't get any prompts to use saved passwords when doing full flow sign-in.

I created an AzureLoginWidget that will pass through to your normal widgets until a full auth flow is required, in
which case it will replace your tree with the webview. I used bloc/flutter_bloc library to support this. When running on flutter
web, the AzureLoginWidget always passes through to the same widget because the state tracking used for mobile is
done entirely in the msal.js library. When using web, the only reliable method of signin I've found has been full redirect that
works across Safari/Chrome browsers; however this means the app is no longer running since it's been entirely replaced with
the Azure auth flow web page until the auth fails or succeeds and redirects back into the app.

What I haven't been able to test and would need effort before this could be considered further:
* anything running on iOS (I currently only have an android build environment available to me)
* anything related to b2c flows (I don't have this scenario)

I'm not sure the simple widget approach for the mobile view will work well with routes and so on as I haven't tested behaviour in those scenarios. I don't have any example code and haven't tried any of these options, but you could try:
* wrapping all scaffold bodies in the routes with the widget
* popping all routes on full flow requirement to the root of the app where the widget is found
* Replacing the current route with a special login route

FiercestT commented:
Hi there, would seriously like to try this out (I need a cross platform Microsoft solution). I add the git package to my pubspec to test it, and get the following
```dart
Launching lib\main.dart on Chrome in debug mode...
lib\main.dart:1
../../../../../../../AppData/Local/Pub/Cache/git/aad_oauth-7b197db139fb6e9990233b7061b97d390eea156b/lib/auth_token_provider.dart:26:7: Error: Method not found: 'getAuthTokenProviderFromConfig'.
      getAuthTokenProviderFromConfig(config);
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Exited (1)
```

I've tried repairing my cache, deleting the git cache manually, even downloading the source and importing the directory as a package and I still get this error. VSCode can't track down the definition for that method either.

stevenebutler commented:
Sorry about that. I had started to add support for constructing via the full AadConfig class (though it doesn't use any of the other options yet), but had missed one of the functions.

I have pushed an update that should help. This update tries to hide some of the differences between mobile and web a bit better though it could still use some work. 

I have run the first included example app on web, and was able to successfully sign in to my tenant, but it doesn't really do much except allow logout/login as MSAL still manages the tokens internally.

I have started trying to unify the mobile/web APIs with this update as the bloc is used for state management for both web and mobile modes, but the web supports only an initial, authenticated and signed out states (and none of the intermediate states).

You must use the AuthTokenProvider interface to login, logout or access the token on the web or you will risk it being out of date. Fortunately, this interface works for mobile as well and you can use Provider package to pass the token provider around the widget tree (see the example). 

Note you should avoid directly adding events to the bloc as this will only be relevant for mobile.

sladomic commented:
Thanks for your contribution. We'll review the PR in the coming weeks and reach back to you!

jacojvv-dev commented:
Is the Android minSdkVersion bump a hard requirement? Would love to take advantage of all of this if it gets merged - but it would mean I need to up the minSdkVersion of my app.

stevenebutler commented:
> Is the Android minSdkVersion bump a hard requirement? Would love to take advantage of all of this if it gets merged - but it would mean I need to up the minSdkVersion of my app.

It may not be required if surfaceandroidwebview isn't used. The widget enables this in Android do I bumped the version. This could be made optional for users though.
From the webview docs:

SurfaceAndroidWebView() requires API level 19. The plugin itself doesn't enforce the API level, so if you want to make the app available on devices running this API level or above, add the following to <your-app>/android/app/build.gradle:

android {
    defaultConfig {
        // Required by the Flutter WebView plugin.
        minSdkVersion 19
    }
}

stevenebutler commented:
> Is the Android minSdkVersion bump a hard requirement? Would love to take advantage of all of this if it gets merged - but it would mean I need to up the minSdkVersion of my app.

I've pushed a commit that removes the minSdkVersion bump. I've left the choice of using hybrid composition to the
library consumer as it should have been, which means the minSdkVersion should be okay to use 18 if this isn't used.

stevenebutler commented:
I'll note that the web support is using implicit flow at the moment, using MSAL v1 js libraries. This should probably be updated to use MSAL v2 to support authorization code flow.

FiercestT commented:
> I'll note that the web support is using implicit flow at the moment, using MSAL v1 js libraries. This should probably be updated to use MSAL v2 to support authorization code flow.

It's very important to move to MSAL 2 and use the auth code grant as the implicit grant will not work unless the client browser has 3rd party cookies enabled. This is a big issue for me using the msal flutter package.

stevenebutler commented:
I think it works when you use redirect login, which this implementation
does, but happy to be told otherwise as maybe I have a different
configuration.

On Sat, 10 Apr 2021, 7:18 pm Fiercest, ***@***.***> wrote:

> I'll note that the web support is using implicit flow at the moment, using
> MSAL v1 js libraries. This should probably be updated to use MSAL v2 to
> support authorization code flow.
>
> It's very important to move to MSAL 2 and use the auth code grant as the
> implicit grant will not work unless the client browser has 3rd party
> cookies enabled. This is a big issue for me using the msal flutter package.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Earlybyte/aad_oauth/pull/103#issuecomment-817106200>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEADFJNQHKPD673TGMRCM5LTIAJX3ANCNFSM4ZQ5FQDA>
> .
>

apply-sindre commented:
@stevenebutler 
Added iOS support in this PR: https://github.com/stevenebutler/aad_oauth/pull/1

apply-sindre commented:
I have one concern, the `login()` and `logout()` methods are actually violating the `Future` contract that the developer expects.

Which makes the following break:
```dart
await azure.logout();
final token = await azure.getAccessToken();
assert(token != null); // token will ALWAYS be [null] due to the timing of the logout state emission
```

I get that it's hard to coordinate this when you have to do `bloc.add(event)`, but one solution would be:
* Make `_LogOutEvent` take a `Completer` as a parameter
* Return the `completer.future` to the caller
* Call `event.completer.complete(null)` from within the BLoC (after the logout is done)

stevenebutler commented:
Thanks for the report. How are you reproducing the issue? Are you using a unit test or running it within an emulator?

stevenebutler commented:
I have tried to fix the future issue but not as you suggested.

Could you try this branch out and let me know if it is working for you now?
```yaml
  aad_oauth:
    git:
      url: https://github.com/stevenebutler/aad_oauth.git
      ref: future-fix
```

stevenebutler commented:
I have added MSAL v2 (authorization code flow) support and updated the documentation. I have left the v1 code in tact, but haven't retested it with the changes made to support v2. Ideally, v1 support will be removed once v2 support has seen some testing and works across devices.

apply-sindre commented:
PERSONAL COMMENT:
I just want to say that I've been using this branch for a long time now, and it's working for my use case at least.
I have been using it both on a single-tenant as well as a multi-tenant application on both iOS and Android.

I would say it's OK to use for production if you're not scared of code, which is kinda the deal with Flutter atm.
You can't really expect these integrations to be AAA "production"-grade.

Vayth commented:
So, is this gonna be merged, or will it be forked into a separated package? I'm not sure where to ask question/issue for this `null-safety` version since it has breaking changes.

stevenebutler commented:
I'd prefer it be merged, with rework if necessary. You can ask questions
about it here as I monitor the issue.

On Tue, 8 June 2021, 6:39 pm Furqana Fathuzzaman, ***@***.***>
wrote:

> So, is this gonna be merged, or will it be forked into a separated
> package? I'm not sure where to ask question/issue for this null-safety
> version since it has breaking changes.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Earlybyte/aad_oauth/pull/103#issuecomment-856578235>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEADFJKXEOXXTJ4AGLMR4WDTRXJNPANCNFSM4ZQ5FQDA>
> .
>

Vayth commented:
Thanks. I'm still trying to understand how to use `AzureLoginWidget`. From what I understand, it's a wrapper for widgets; when `login()` is called, it returns AAD webview. Otherwise, it returns the specified widget.

Let say I have a `LoginPage` with logo and a button which call `login()`.
On previous version, I'd call `login()` inside a `try` in my `OldAuthProvider` (which extends a `ChangeNotifier`).
Then in my `LoginPage`, I did `Navigator.pushReplacement` to my `HomePage`

In this new one, how should I approach it? Do I use AzureLoginWidget as a wrapper for my whole application?
Since I'm not sure where to put `Navigator`.
Do I use it like this?
```
AzureLoginWidget(
  authTokenProvider: tokenProvider,
  whenAuthenticated: HomePage(),
  whenInitial: LoginPage(),
  whenLoginFailed: LoginPage('Login Failed'),
  whenSignedOut: LoginPage('Signed Out'),
```

stevenebutler commented:
The login widget is designed to show a different view based on the bloc
state.

If you want to move to a different route to login you may need to use a
different approach as I haven’t really catered for that scenario. The login
widget would normally sit in the scaffold at the body level and wrap your
normal body widget, replacing it with the webview if it needs to login and
showing your widget when logged in.

Some feedback on better approaches to work with navigation is one of the
things I had hoped I’d get by submitting the pull request.

It should be possible to wrap any route’s scaffold with the login widget
but that might not be the nicest approach.

An approach that handles routes could define a widget that takes a callback
for successful login that changes the route, instead of wrapping the ui but
that would also require something to navigate back to the login route any
time auth was lost.

On Tue, 8 Jun 2021 at 8:36 pm, Furqana Fathuzzaman ***@***.***>
wrote:

> Thanks. I'm still trying to understand how to use AzureLoginWidget. From
> what I understand, it's a wrapper for widgets; when login() is called, it
> returns AAD webview. Otherwise, it returns the specified widget.
>
> Let say I have a LoginPage with logo and a button which call login().
> On previous version, I'd call login() inside a try in my OldAuthProvider
> (which extends a ChangeNotifier).
> Then in my LoginPage, I did Navigator.pushReplacement to my HomePage
>
> In this new one, how should I approach it? Do I use AzureLoginWidget as a
> wrapper for my whole application?
> Since I'm not sure where to put Navigator.
> Do I use it like this?
>
> AzureLoginWidget(
>   authTokenProvider: tokenProvider,
>   whenAuthenticated: HomePage(),
>   whenInitial: LoginPage(),
>   whenLoginFailed: LoginPage('Login Failed'),
>   whenSignedOut: LoginPage('Signed Out'),
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Earlybyte/aad_oauth/pull/103#issuecomment-856653628>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEADFJPUHEMQBYEHIIHNTQ3TRXXEHANCNFSM4ZQ5FQDA>
> .
>
-- 
Regards,
Steven Butler

sladomic commented:
Hey @stevenebutler thanks again for the amazing work. I was trying to get it running in both the example and our real app. Example seems to work, but I have a hard time getting it running in our real app, since a lot of package dependencies have been added here which give me some version conflicts (e. g. flutter_sentry). Is it possible to have less strict package version requirements in your pubspec?

stevenebutler commented:
Some of these might come from other packages like bloc. I will have to take
a look but most of the packages I used are the minimum version that support
null safety which is an explicit goal is the patch.

On Wed, 9 June 2021, 4:24 pm sladomic, ***@***.***> wrote:

> Hey @stevenebutler <https://github.com/stevenebutler> thanks again for
> the amazing work. I was trying to get it running in both the example and
> our real app. Example seems to work, but I have a hard time getting it
> running in our real app, since a lot of package dependencies have been
> added here which give me some version conflicts (e. g. flutter_sentry). Is
> it possible to have less strict package version requirements in your
> pubspec?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Earlybyte/aad_oauth/pull/103#issuecomment-857417464>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEADFJLRMCYIYWE6L7S7CWTTR4CI5ANCNFSM4ZQ5FQDA>
> .
>

sladomic commented:
> Some of these might come from other packages like bloc. I will have to take a look but most of the packages I used are the minimum version that support null safety which is an explicit goal is the patch.
> […](#)
> On Wed, 9 June 2021, 4:24 pm sladomic, ***@***.***> wrote: Hey @stevenebutler <https://github.com/stevenebutler> thanks again for the amazing work. I was trying to get it running in both the example and our real app. Example seems to work, but I have a hard time getting it running in our real app, since a lot of package dependencies have been added here which give me some version conflicts (e. g. flutter_sentry). Is it possible to have less strict package version requirements in your pubspec? — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <[#103 (comment)](https://github.com/Earlybyte/aad_oauth/pull/103#issuecomment-857417464)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AEADFJLRMCYIYWE6L7S7CWTTR4CI5ANCNFSM4ZQ5FQDA> .

makes sense, so the other packages are just not up-to-date as well :) I'd like to do a dev release so other people here can start using it, but we might need to wait on the other packages for us to be able to use it

stevenebutler commented:
What versions are you looking for and what versions is the package forcing
on you?

On Wed, 9 June 2021, 5:33 pm sladomic, ***@***.***> wrote:

> Some of these might come from other packages like bloc. I will have to
> take a look but most of the packages I used are the minimum version that
> support null safety which is an explicit goal is the patch.
> … <#m_-8974300007240299004_>
> On Wed, 9 June 2021, 4:24 pm sladomic, *@*.***> wrote: Hey @stevenebutler
> <https://github.com/stevenebutler> https://github.com/stevenebutler
> thanks again for the amazing work. I was trying to get it running in both
> the example and our real app. Example seems to work, but I have a hard time
> getting it running in our real app, since a lot of package dependencies
> have been added here which give me some version conflicts (e. g.
> flutter_sentry). Is it possible to have less strict package version
> requirements in your pubspec? — You are receiving this because you were
> mentioned. Reply to this email directly, view it on GitHub <#103 (comment)
> <https://github.com/Earlybyte/aad_oauth/pull/103#issuecomment-857417464>>,
> or unsubscribe
> https://github.com/notifications/unsubscribe-auth/AEADFJLRMCYIYWE6L7S7CWTTR4CI5ANCNFSM4ZQ5FQDA
> .
>
> makes sense, so the other packages are just not up-to-date as well :) I'd
> like to do a dev release so other people here can start using it, but we
> might need to wait on the other packages for us to be able to use it
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Earlybyte/aad_oauth/pull/103#issuecomment-857460961>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEADFJPNROVYREV4IKIBMSDTR4KMZANCNFSM4ZQ5FQDA>
> .
>

stevenebutler commented:
I don't think flutter_sentry is in the dependency chain for this package at
all (it's not listed in the pubspec.lock file anyway). Is the problem that
this package depends on a version of a package that flutter_sentry also
depends on, and they are conflicting incompatible versions?

In any case I've pushed an update that bumps all package versions to latest
stable versions available, so hopefully that will help if it's pulling in
an older dependency.



On Wed, 9 Jun 2021 at 17:40, Steven Butler ***@***.***> wrote:

> What versions are you looking for and what versions is the package forcing
> on you?
>
> On Wed, 9 June 2021, 5:33 pm sladomic, ***@***.***> wrote:
>
>> Some of these might come from other packages like bloc. I will have to
>> take a look but most of the packages I used are the minimum version that
>> support null safety which is an explicit goal is the patch.
>> … <#m_-4931174561494233235_m_-8974300007240299004_>
>> On Wed, 9 June 2021, 4:24 pm sladomic, *@*.***> wrote: Hey @stevenebutler
>> <https://github.com/stevenebutler> https://github.com/stevenebutler
>> thanks again for the amazing work. I was trying to get it running in both
>> the example and our real app. Example seems to work, but I have a hard time
>> getting it running in our real app, since a lot of package dependencies
>> have been added here which give me some version conflicts (e. g.
>> flutter_sentry). Is it possible to have less strict package version
>> requirements in your pubspec? — You are receiving this because you were
>> mentioned. Reply to this email directly, view it on GitHub <#103
>> (comment)
>> <https://github.com/Earlybyte/aad_oauth/pull/103#issuecomment-857417464>>,
>> or unsubscribe
>> https://github.com/notifications/unsubscribe-auth/AEADFJLRMCYIYWE6L7S7CWTTR4CI5ANCNFSM4ZQ5FQDA
>> .
>>
>> makes sense, so the other packages are just not up-to-date as well :) I'd
>> like to do a dev release so other people here can start using it, but we
>> might need to wait on the other packages for us to be able to use it
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/Earlybyte/aad_oauth/pull/103#issuecomment-857460961>,
>> or unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/AEADFJPNROVYREV4IKIBMSDTR4KMZANCNFSM4ZQ5FQDA>
>> .
>>
>

-- 
Regards,
Steven Butler

rajeshtaneja commented:
Can someone please review it ?

sladomic commented:
Sorry for the long wait. Still reviewing since it's a large and breaking change. I've resolved all the package problems in our productive app so I can test it there as well. What I found already:
- B2C example not updated
- Login persists across reinstalls

Details for the 2nd point:
#### Description
1. Install and login
2. Uninstall
3. Install and login
#### Expected
User has to enter credentials again, because the token isn't persisted across reinstalls

#### Actual
User doesn't have to enter credentials again because token is still valid

alepalermo97 commented:
Hi @stevenebutler, thank you for the amazing work. I tested the package on iOS/Android/web and everything works fine. It would be nice to have a method to expose the entire Token instead of only the accessToken (maybe leave _getAccesToken()_ and add _getToken()_). In my usecase, for example, I needed the _idToken_ instead of _accessToken_, so I had to make a change in the pachage's code to make it work. Btw, I don't know if there is actually a side effect of adding it. Thank you again

stevenebutler commented:
I couldn't find a way to get the full token or of the web js libraries
using the approved API. It could be done by using storage directly though I
think.

Were you able to do this?

I know sladomic had some review comments on the change I've yet to address
as I've not had the time lately to look at it.

On Mon, 28 June 2021, 6:55 pm alepalermo97, ***@***.***>
wrote:

> Hi @stevenebutler <https://github.com/stevenebutler>, thank you for the
> amazing work. I tested the package on iOS/Android/web and everything works
> fine. It would be nice to have a method to expose the entire Token instead
> of only the accessToken (maybe leave *getAccesToken()* and add
> *getToken()*). In my usecase, for example, I needed the *idToken* instead
> of *accessToken*, so I had to make a change in the pachage's code to make
> it work. Btw, I don't know if there is actually a side effect of adding it.
> Thank you again
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Earlybyte/aad_oauth/pull/103#issuecomment-869502718>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEADFJIWRMZMRVKKWUWEBA3TVA2H5ANCNFSM4ZQ5FQDA>
> .
>

stevenebutler commented:
> Sorry for the long wait. Still reviewing since it's a large and breaking change. I've resolved all the package problems in our productive app so I can test it there as well. What I found already:
> 
> * B2C example not updated (probably @erickdaros or @jacojvv-dev can help)
> * Login persists across reinstalls
> 
> Details for the 2nd point:
> 
> #### Description
> 1. Install and login
> 2. Uninstall
> 3. Install and login
> 
> #### Expected
> User has to enter credentials again, because the token isn't persisted across reinstalls
> 
> #### Actual
> User doesn't have to enter credentials again because token is still valid

Hi @sladomic, thanks for the review and feedback. I hadn't considered the need for clearing the tokens on reinstall an issue, as normally this would be installed under the user's account. Could you explain the reasons behind the requirement to me?

I had initially thought it was to handle the secure key being unable to encrypt the token left behind when changing devices, and thought I'd handled that through the retry logic, but if there are other reasons behind it I'm all ears.

I definitely would appreciate help with the B2C example, as I mentioned in the OP I have no use-case for those features in my environment and would be unable to test they are working.

I also have to play my own devil's advocate and wonder that if now the webview_flutter_plugin has support for null safety in 0.4.0 it wouldn't be a simpler upgrade if the work I've done here was instead done more in-line with the current implementation. 

This might help with easing migration pain as the APIs may not need to change so drastically to support web (I would hope). The major driver for the changes I made were to support the in-widget-tree flutter webview component.

What are your thoughts on the pros/cons of migrating to the standard webview vs the webview_flutter_plugin which has a notice of intent to deprecate in the future?

stevenebutler commented:
> Sorry for the long wait. Still reviewing since it's a large and breaking change. I've resolved all the package problems in our productive app so I can test it there as well. What I found already:
> 
> * B2C example not updated (probably @erickdaros or @jacojvv-dev can help)
> * Login persists across reinstalls
> 
> Details for the 2nd point:
> 
> #### Description
> 1. Install and login
> 2. Uninstall
> 3. Install and login
> 
> #### Expected
> User has to enter credentials again, because the token isn't persisted across reinstalls
> 
> #### Actual
> User doesn't have to enter credentials again because token is still valid

Just to clarify this issue report, are you installing the app through the play/apple store, or side loading through adb or ios equivalent? Just to reiterate, I don't have an iOS development environment so can't test any changes I make in that environment.

stevenebutler commented:
Latest version of the webview_plugin requires minSdk of 19 (possibly because SurfaceAndroidWebView requires that), so I'm going to have to bump the minSdk of this patch up to 19 as well.

sladomic commented:
> webview_flutter_plugin

It just seems best practice to me and I think most users expect an app to not persist login data after a fresh install. I just uninstalled Airbnb from my device and reinstalled it again and there the login isn't persisted as well.

Regarding flutter_webview_plugin I'm totally with you and think it's best to move to the standard webview.

I'll make a release 0.2.2 today with updated packages, so people are not forced to use your version with breaking changes right away, but I'd like to use your version as a release 0.3.0 or 1.0.0.

I still have some issues getting it running in my clients app, but I'll figure it out.

stevenebutler commented:
> > webview_flutter_plugin
> 
> It just seems best practice to me and I think most users expect an app to not persist login data after a fresh install. I just uninstalled Airbnb from my device and reinstalled it again and there the login isn't persisted as well.

That's fair, but when you tested was that on iOS or Android? As I mentioned, the old key doesn't work on my android device due to the flutter secure storage private key changing on reinstall, and being unable to decrypt the old one, so it's hard to test I've got it right with my setup. I assume the shared preference logic is there to detect when the app is run for the first time and remove the old key before the normal logic tries to read it.

> Regarding flutter_webview_plugin I'm totally with you and think it's best to move to the standard webview.
> 
> I'll make a release 0.2.2 today with updated packages, so people are not forced to use your version with breaking changes right away, but I'd like to use your version as a release 0.3.0 or 1.0.0.
> 
> I still have some issues getting it running in my clients app, but I'll figure it out.

Ok, that's cool. I just wanted to make sure you're aware of the drawbacks to the official webview plugin that I mentioned in the OP. 

* Android keyboard doesn't suggest passwords any more (works with the original). I'm not sure if this affects iOS or not.
* Managing a login widget into the scaffold when most apps would have multiple routes. 
  * The original webview can float above everything regardless of active route when login is lost. The replacement widget doesn't have any logic to deal with pushed routes etc, so may not show up in some cases. In practice it has worked in my app, but that's because it always opens on the home route with the login widget wrapping the body of the scaffold. I'm not sure what would be necessary if the app and refresh token expired while deep in the app, because in practice this just doesn't happen (when you request an offline token).
  * Having said that, users with need to do advanced routing don't have to use the supplied widget, and could provide their own, using the bloc state to drive it.

stevenebutler commented:
> > webview_flutter_plugin
> 
> It just seems best practice to me and I think most users expect an app to not persist login data after a fresh install. I just uninstalled Airbnb from my device and reinstalled it again and there the login isn't persisted as well.

Commit 309092b adds the shared_preferences code back into the cache handler that tries to remove the token on first login after reinstall. Could you retest the reinstall scenario, as it still requires me to login each time on my device after reinstall, but it always worked that way on my device.

sladomic commented:
> > > webview_flutter_plugin
> > 
> > 
> > It just seems best practice to me and I think most users expect an app to not persist login data after a fresh install. I just uninstalled Airbnb from my device and reinstalled it again and there the login isn't persisted as well.
> 
> Commit [309092b](https://github.com/Earlybyte/aad_oauth/commit/309092b45713ae1f69f87105ba64817128a9071c) adds the shared_preferences code back into the cache handler that tries to remove the token on first login after reinstall. Could you retest the reinstall scenario, as it still requires me to login each time on my device after reinstall, but it always worked that way on my device.

seems to be an [iOS-related issue](https://github.com/mogol/flutter_secure_storage/issues/88)

works as expected again :)

stevenebutler commented:
I've updated according to latest release though most of the changes were discarded.

I've noticed that in the current release the getAccessToken() no longer refreshes automatically, but in this PR, it still does, and login still just does a getAccessToken() request. This PR also doesn't support the refreshIfAvailable flag on login.

I'm not sure why anyone would want to get a stale access token though, so I'm curious as to why this change was made, or am I missing something?

chester23491 commented:
We discussed this pull request in earlybyte. It's quite complex change, so the pull request is unreviewable.
We propose the following strategy to achieve the changes into the project.

1. Implement null safety on top of existing solution (Issue: #105) 
2. Replace `flutter_webview_plugin` with `webview_flutter`
3. Add flutter web support (Issue: #106) 

Our main goal is to avoid breaking changes with this strategy. We like to discuss this with the community and are planning meeting on 23.07.2021 16:00 (UTC+2). Please fill out the [registration form](https://teams.microsoft.com/registration/myhG37bBCkGephSLjdMQng,C7IIcyr1BEucSJ8171UJUA,jjw0TxXyG0ybv6K8uE4uTw,f5TxUMIByUyT0HOHMtX6Rg,FogZR4NAOECaYQjhKWS1PA,kr6RXtZL-Ua1QH_XDigV3w?mode=read&tenantId=df46289b-c1b6-410a-9ea6-148b8dd3109e) to join the meeting.

Thank you for your contribution and looking forward to meet.

stevenebutler commented:
Thanks for your response. I won't be able to make that meeting as it will
start at midnight. My only suggestion would be to reverse the order of
points 2 and 3, assuming the work I did to introduce web support in this PR
is used as a basis of the support. It actually aligns better with the
existing flutter_webview_plugin model than it does with the webview_flutter
model for webviews.


On Thu, 15 Jul 2021 at 18:17, chester23491 ***@***.***> wrote:

> We discussed this pull request in earlybyte. It's quite complex change, so
> the pull request is unreviewable.
> We propose the following strategy to achieve the changes into the project.
>
>    1. Implement null safety on top of existing solution (Issue: #105
>    <https://github.com/Earlybyte/aad_oauth/issues/105>)
>    2. Replace flutter_webview_plugin with webview_flutter
>    3. Add flutter web support (Issue: #106
>    <https://github.com/Earlybyte/aad_oauth/issues/106>)
>
> Our main goal is to avoid breaking changes with this strategy. We like to
> discuss this with the community and are planning meeting on 23.07.2021
> 16:00 (UTC+2). Please fill out the registration form
> <https://teams.microsoft.com/registration/myhG37bBCkGephSLjdMQng,C7IIcyr1BEucSJ8171UJUA,jjw0TxXyG0ybv6K8uE4uTw,f5TxUMIByUyT0HOHMtX6Rg,FogZR4NAOECaYQjhKWS1PA,kr6RXtZL-Ua1QH_XDigV3w?mode=read&tenantId=df46289b-c1b6-410a-9ea6-148b8dd3109e>
> to join the meeting.
>
> Thank you for your contribution and looking forward to meet.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Earlybyte/aad_oauth/pull/103#issuecomment-880495281>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEADFJOZ4F6PUJE5Y7CWBYLTX2KP3ANCNFSM4ZQ5FQDA>
> .
>


-- 
Regards,
Steven Butler

chester23491 commented:
@stevenebutler - Regarding the time, we can also schedule earlier to 08:30 (UTC+2)?

stevenebutler commented:
Thanks, I'm more likely able to make that time.

On Thu, 15 Jul 2021 at 19:46, chester23491 ***@***.***> wrote:

> @stevenebutler <https://github.com/stevenebutler> - Regarding the time,
> we can also schedule earlier to 08:30 (UTC+2)?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Earlybyte/aad_oauth/pull/103#issuecomment-880555468>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEADFJL5FJA75ZCFX7ZP4ZLTX2U7LANCNFSM4ZQ5FQDA>
> .
>


-- 
Regards,
Steven Butler

chester23491 commented:
The time of the meeting was rescheduled to 08:30am (UTC+2), the registration form is updated and you can check the timezone offsets [here](https://www.timeanddate.com/worldclock/fixedtime.html?msg=Community+Meeting&iso=20210723T0630&p1=1440&ah=1).

stevenebutler commented:
Following discussion with the earlybyte maintainers, I am withdrawing this PR, though I'll leave the branch active for anyone using it until such time the features in it are supported by the main package.
